File: Java/ScrambleString.java
public class ScrambleString {
    
    //DP solution, O(n^4) time and O(n^3) space
    //b[l][i1][i2] is true if s1[i1...i1+l-1] is a scrambled string of s2[i2...i2+l-1]
    public boolean isScramble0(String s1, String s2) {
        int len = s1.length();
        if (s2.length() != len) {
            return false;
        } else if (len == 0) {
            return true;
        }
        boolean[][][] b = new boolean[len + 1][len][len];
        for (int i = 0; i < len; ++i) {
            for (int j = 0; j < len; ++j) {
                b[1][i][j] = s1.charAt(i) == s2.charAt(j);
            }
        }
        for (int l = 2; l <= len; ++l) {
            for (int i1 = 0; i1 <= len - l; ++i1) {
                for (int i2 = 0; i2 <= len - l; ++i2) {
                    b[l][i1][i2] = false;
                    for (int lx = 1; lx < l; ++lx) {
                        if ((b[lx][i1][i2] && b[l-lx][i1+lx][i2+lx]
                         || (b[lx][i1][i2+l-lx] && b[l-lx][i1+lx][i2]))) {
                             b[l][i1][i2] = true;
                             break;
                         }
                    }
                }
            }
        }
        return b[len][0][0];
    }
    
    private boolean fastNo(String s1, String s2, int i1, int i2, int len) {
        Map<Character, Integer> counts = new HashMap<>();
        for (int i = i1; i < i1+len; ++i) {
            int preCount = counts.getOrDefault(s1.charAt(i), 0);
            counts.put(s1.charAt(i), ++preCount);
        }
        for (int i = i2; i < i2+len; ++i) {
            int preCount = counts.getOrDefault(s2.charAt(i), 0);
            if (preCount == 0) {
                return false;
            }
            counts.put(s2.charAt(i), --preCount);                
        }
        return true;
    }
    
    private boolean isScrambleRecursive(String s1, String s2, int i1, int i2, int len, Boolean[][][] cache) {
        if (cache[len][i1][i2] != null) {
            return cache[len][i1][i2];
        }
        if (len == 1) {
            cache[len][i1][i2] = s1.charAt(i1) == s2.charAt(i2);
        } else {
            cache[len][i1][i2] = false;
            if (!fastNo(s1, s2, i1, i2, len)) {
                return false;
            }
            for (int lx = 1; lx < len; ++lx) {
                if ((isScrambleRecursive(s1, s2, i1, i2, lx, cache) 
                    && isScrambleRecursive(s1, s2, i1+lx, i2+lx, len-lx, cache))
                 || (isScrambleRecursive(s1, s2, i1, i2+len-lx, lx, cache)
                    && isScrambleRecursive(s1, s2, i1+lx, i2, len-lx, cache))) {
                        cache[len][i1][i2] = true;
                        break;
                    }
            }
        }
        return cache[len][i1][i2];
    }
    
    //Faster solution using Memoization
    public boolean isScramble(String s1, String s2) {
        int len = s1.length();
        if (s2.length() != len) {
            return false;
        } else if (len == 0) {
            return true;
        }
        Boolean[][][] cache = new Boolean[len + 1][len][len];
        return isScrambleRecursive(s1, s2, 0, 0, len, cache);
    }
}

File: C++/ScrambleString.cc
#include <iostream>
#include <vector>
using namespace std;

//DP solution
//O(n^4) time and O(n^3) space
bool isScramble0(string s1, string s2) {
    if (s1.size() != s2.size())
        return false;

    int n = s1.size();

    vector<vector<vector<bool> > > f(n+1, vector<vector<bool> >(n, vector<bool>(n, false)));
   
    //These two lines are much faster than the above
  /*  bool f[n + 1][n][n];
    fill_n(&f[0][0][0], (n + 1) * n * n, false);*/

    for (int i = 0; i < n; ++i)
        for (int j = 0; j < n; ++j)
            f[1][i][j] = (s1[i] == s2[j]);

    for (int l = 2; l <= n; ++l)
    {
        for (int i = 0; i+l <= n; ++i)
        {
            for (int j = 0; j+l <= n; ++j)
            {
                for (int t = 1; t < l; ++t)
                {
                    if ((f[t][i][j] && f[l-t][i+t][j+t]) ||
                            (f[t][i][j+l-t] && f[l-t][i+t][j])) {
                        
                        f[l][i][j] = true;
                        break;
                    }

                }
            }
        }
    }

    return f[n][0][0];
}

//Improved implementation
bool isScramble(string s1, string s2) {
    int n1 = s1.size();
    int n2 = s2.size();
    if (n1 != n2)
        return false;
    
    vector<vector<vector<bool> > > cache(n1+1, vector<vector<bool> >(n1, vector<bool>(n1, false)));
    
    for (int l = 1; l <= n1; ++l)
    {
        for (int i = 0; i+l <= n1; ++i)
        {
            for (int j = 0; j+l <= n1; ++j)
            {
                if (l == 1)
                    cache[l][i][j] = (s1[i] == s2[j]);
                    
                for (int subl = 1; subl < l; ++subl)
                {
                    if ((cache[subl][i][j] && cache[l-subl][i+subl][j+subl])
                    || (cache[subl][i][j+l-subl] && cache[l-subl][i+subl][j])) {
                        cache[l][i][j] = true;
                        break;
                    }
                }
            }
        }
    }
    
    return (n1 == 0) ? true : cache[n1][0][0];
}


int main(int argc, char** argv)
{
    string s1 = "great";
    string s2 = "rgtae";
    cout << isScramble(s1, s2) << endl; //Should be true
    return 0;
}

